amethyst_input/src/bindings.rs

```diff
@@ -47,10 +49,90 @@ where
     pub(super) actions: HashMap<AC, SmallVec<[SmallVec<[Button; 2]>; 4]>>,
 }
 
+/// An enum of possible errors that can occur when binding an action or axis.
+#[derive(Debug, Clone, PartialEq)]
+pub enum BindingError<AX, AC> {
+    /// Combo provided for action binding has two (or more) of the same button.
+    ComboContainsDuplicates,
+    /// Combo provided was already bound to contained action.
+    ComboAlreadyBound(AC),
+    /// A combo of length 1 was provided, and it overlaps with an axis binding.
+    ButtonBoundToAxis(AX, Axis),
+    /// Axis buttons provided have overlap with an existing axis
+    AxisButtonAlreadyBoundToAxis(AX, Axis),
+    /// Axis buttons have overlap with an action combo of length 1.
+    AxisButtonAlreadyBoundToAction(AC, Button),
+    /// That specific axis on that specific controller is already in use for an
+    /// axis binding.
+    ControllerAxisAlreadyBound(AX),
+}
+
+impl<AX, AC> Display for BindingError<AX, AC> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match *self {
+            BindingError::ComboContainsDuplicates => write!(
+                f,
+                "Combo provided contained two (or more) of the same button"
+            ),
+            BindingError::ComboAlreadyBound(ref _action) => {
```

Comment:
I'm not super happy about the fact that I can't use `AX` and `AC` debug representations here without requiring `Debug` for `AX` and `AC`. I don't think it'd be that big of a deal to require `Debug` though. I'm curious what others think, specifically @azriel91 as I know you use this with non-String types.